crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
     }
 }
 
+#[derive(Debug)]
+enum StringQuoteChar {
+    Single,
+    Double,
+}
+
+impl TryFrom<char> for StringQuoteChar {
+    type Error = String;
+
+    fn try_from(c: char) -> Result<Self, Self::Error> {
+        match c {
+            '\'' => Ok(Self::Single),
+            '"' => Ok(Self::Double),
+            _ => Err(format!("Invalid string quote character: {c}")),
+        }
+    }
+}
+
+impl StringQuoteChar {
+    const fn as_char(&self) -> char {
+        match self {
+            Self::Single => '\'',
+            Self::Double => '"',
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StringQuoteSize {
+    Single,
+    Triple,
+}
+
+#[derive(Debug)]
+struct FStringContext {
+    quote_char: StringQuoteChar,
+    quote_size: StringQuoteSize,
+    parentheses_count: u32,
+    format_spec_count: u32,
+    is_raw_string: bool,
```

Comment:
Nit: We could potentially squeeze `quote_char`, `quote_size` and `raw` (3 bytes) into a single byte by using bitflags. It won't give us any space-saving because of padding but may be easier to expand.

You would have three flags:

* TRIPLE (missing implies single)
* DOUBLE (missing implies single)
* RAW (missing implies non-raw)

Reply:
This is a good idea! I've implemented it and it makes the code which uses the `FStringContext` better to reason about.